contracts/farm/LiquidityMining.sol

```diff
@@ -0,0 +1,567 @@
+		_transferReward(user, userAddress, false);
+
+		user.amount = user.amount.sub(_amount);
+		pool.poolToken.safeTransfer(address(msg.sender), _amount); //sent to the user or wrapper
+		_updateRewardDebt(pool, user);
+		emit Withdraw(userAddress, _poolToken, _amount);
+	}
+
+	function _getUserAddress(address _user) internal view returns (address) {
+		address userAddress = msg.sender;
+		if (_user != address(0)) {
+			//only wrapper can pass _user parameter
+			require(msg.sender == wrapper && _user == tx.origin, "unauthorized");
+			userAddress = _user;
+		}
+		return userAddress;
+	}
+
+	function _updateReward(PoolInfo storage pool, UserInfo storage user) internal {
+		//update user accumulated reward
+		if (user.amount > 0) {
+			//add reward for the previous amount of deposited tokens
+			uint256 accumulatedReward = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION).sub(user.rewardDebt);
+			user.accumulatedReward = user.accumulatedReward.add(accumulatedReward);
+		}
+	}
+
+	function _updateRewardDebt(PoolInfo storage pool, UserInfo storage user) internal {
+		//reward accumulated before amount update (should be subtracted during next reward calculation)
+		user.rewardDebt = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION);
+	}
+
+	/**
+	 * @notice Send reward in SOV to the lockedSOV vault.
+	 * @param _user The user info, to get its reward share.
+	 * @param _userAddress The address of the user, to send SOV in its behalf.
+	 */
+	function _transferReward(
+		UserInfo storage _user,
+		address _userAddress,
+		bool _isClaimingReward
+	) internal {
+		uint256 userAccumulatedReward = _user.accumulatedReward;
+
+		/// @dev Transfer if enough SOV balance on this LM contract.
+		uint256 balance = SOV.balanceOf(address(this));
+		if (balance >= userAccumulatedReward) {
+			totalUsersBalance = totalUsersBalance.sub(userAccumulatedReward);
+			_user.accumulatedReward = 0;
+
+			/// @dev Instead of transferring the reward to the LP (user),
+			///   deposit it into lockedSOV vault contract, but first
+			///   SOV deposit must be approved to move the SOV tokens
+			///   from this LM contract into the lockedSOV vault.
+			SOV.approve(address(lockedSOV), userAccumulatedReward);
+			lockedSOV.deposit(_userAddress, userAccumulatedReward, basisPoint);
+
+			if (_isClaimingReward) {
+				lockedSOV.withdrawAndStakeTokensFrom(_userAddress);
+			}
+
+			/// @dev Event log.
+			emit RewardClaimed(_userAddress, userAccumulatedReward);
+		}
+	}
+
+	/**
+	 * @notice withdraws pool tokens without transferring reward tokens
+	 * @param _poolToken the address of pool token
+	 * @dev EMERGENCY ONLY
+	 */
+	function emergencyWithdraw(address _poolToken) public {
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][msg.sender];
+
+		uint256 userAmount = user.amount;
+		user.amount = 0;
+		user.rewardDebt = 0;
+		user.accumulatedReward = 0;
```

Comment:
Are there totals that need to be updated? e.g. totalUserBalance?

Reply:
`struct UserInfo {
		uint256 amount; // How many pool tokens the user has provided.
		uint256 rewardDebt; // Reward debt. See explanation below.
		uint256 accumulatedReward; //Reward that's ready to be transferred
}`

All fields will be 0
We don't have any other state variables related to the user.